File: pipeline/strokes.py
"""Anisotropic stroke field via Line Integral Convolution.

Given a per-pixel orientation field (theta, coherence), generate a noise-like
field that is "combed" along the local tangent direction. Where coherence is
low, smoothly blend back to isotropic noise.

Pipeline per pixel (vectorized via scipy.ndimage.map_coordinates):
    1. Sample white noise along the streamline (cos θ, sin θ) at integer
       offsets in [-length/2, +length/2]; average -> directional smoothing.
    2. Sample the result perpendicular to the streamline at integer offsets in
       [-thickness/2, +thickness/2] with Gaussian weights -> thickness blur.
    3. Blend the streamlined output with isotropic noise using coherence.

Output is centered around 0 and roughly bounded in [-1, 1].
"""

from __future__ import annotations

import numpy as np
from scipy.ndimage import map_coordinates


def _normalize_to_unit_amplitude(arr: np.ndarray) -> np.ndarray:
    """Scale so that 99th percentile of |x| ≈ 1; clip to [-1, 1]."""
    p99 = float(np.percentile(np.abs(arr), 99))
    if p99 < 1e-8:
        return arr.astype(np.float32)
    out = arr / p99
    return np.clip(out, -1.0, 1.0).astype(np.float32)


def directional_stroke_field(
    theta: np.ndarray,
    coherence: np.ndarray,
    length: float,
    thickness: float,
    seed: int = 0,
    direction_strength: float = 1.0,
    iso_weight: float = 1.0,
) -> np.ndarray:
    """Generate an anisotropic noise field aligned to the orientation.

    Parameters
    ----------
    theta : np.ndarray (H, W)
        Tangent orientation in radians (any range; only direction matters).
    coherence : np.ndarray (H, W)
        Per-pixel anisotropy strength in [0, 1].
    length : float
        Streamline integration length in pixels (full width). 0 -> isotropic.
    thickness : float
        Perpendicular Gaussian blur width in pixels. 0 -> no thickness blur.
    seed : int
        RNG seed for the underlying white noise (deterministic outputs).
    direction_strength : float
        Multiplier on coherence used for the blend (0..1+). 1.0 = use coherence
        directly. Higher = stroke more visible in low-coherence regions.

    Returns
    -------
    np.ndarray of float32 (H, W) in roughly [-1, 1], mean ≈ 0.
    """
    if theta.shape != coherence.shape:
        raise ValueError(
            f"theta {theta.shape} and coherence {coherence.shape} must match"
        )
    if theta.ndim != 2:
        raise ValueError(f"theta must be 2D, got {theta.shape}")

    H, W = theta.shape
    rng = np.random.default_rng(seed)
    noise = rng.standard_normal((H, W)).astype(np.float32)

    # Coordinate grid (y, x).
    yy, xx = np.meshgrid(
        np.arange(H, dtype=np.float32),
        np.arange(W, dtype=np.float32),
        indexing="ij",
    )

    dx = np.cos(theta).astype(np.float32)
    dy = np.sin(theta).astype(np.float32)

    # ---- Step 1: streamline integration (line integral convolution) ----
    if length >= 1.0:
        n_steps = max(3, int(round(length)))
        if n_steps % 2 == 0:
            n_steps += 1
        half = (n_steps - 1) // 2

        acc = np.zeros((H, W), dtype=np.float32)
        for k in range(-half, half + 1):
            sample_y = yy + k * dy
            sample_x = xx + k * dx
            sampled = map_coordinates(
                noise,
                [sample_y, sample_x],
                order=1,
                mode="reflect",
            )
            acc += sampled.astype(np.float32)
        stroked = acc / float(n_steps)
    else:
        stroked = noise.copy()

    # ---- Step 2: perpendicular Gaussian thickness blur ----
    if thickness >= 1.0:
        n_thick = max(3, int(round(thickness)))
        if n_thick % 2 == 0:
            n_thick += 1
        thalf = (n_thick - 1) // 2
        sigma_t = max(thickness / 2.0, 0.5)
        ks = np.arange(-thalf, thalf + 1, dtype=np.float32)
        weights = np.exp(-0.5 * (ks / sigma_t) ** 2)
        weights = weights / weights.sum()

        acc2 = np.zeros_like(stroked)
        # Perpendicular direction: rotate 90° -> (-sin θ, cos θ) i.e. (-dy, dx)
        for k, w in zip(ks, weights):
            sample_y = yy + k * dx       # perp y component is +cos θ = +dx
            sample_x = xx + k * (-dy)    # perp x component is -sin θ = -dy
            sampled = map_coordinates(
                stroked,
                [sample_y, sample_x],
                order=1,
                mode="reflect",
            )
            acc2 += sampled.astype(np.float32) * w
        stroked = acc2

    # ---- Step 3: blend with isotropic noise via coherence ----
    iso = noise  # already shares the same RNG seed

    # Normalize each component independently so they live on a comparable scale.
    stroked_n = _normalize_to_unit_amplitude(stroked)
    iso_n = _normalize_to_unit_amplitude(iso)

    blend = np.clip(coherence.astype(np.float32) * float(direction_strength), 0.0, 1.0)
    iw = float(iso_weight)
    out = blend * stroked_n + (1.0 - blend) * iso_n * iw
    return out.astype(np.float32)

File: web/src/strokes.js
// 各向异性 stroke field（Line Integral Convolution）。
//
// 移植自 pipeline/strokes.py。算法：
//   1. 白噪声 noise(H, W) = N(0, 1)。
//   2. 沿切线方向 (cos θ, sin θ) 在 [-length/2, +length/2] 取整数步长采样并平均
//      → 平滑出 "笔触" 形状。
//   3. 沿垂直方向做 Gaussian 加权采样 → 笔触厚度。
//   4. 与各向同性噪声按 coherence 做混合：低 coherence 区域回退到平滑。
//
// 性能要点：把双线性采样 inline 进双层循环，避免每个 step 分配 O(N) 临时数组。

import { standardNormalArray } from './rng.js';
import { percentileAbs } from './filters.js';

/* BORDER_REFLECT_101 索引 (重复一份，避免跨模块函数调用的开销) */
function reflect101(i, n) {
    if (n <= 1) return 0;
    const period = 2 * (n - 1);
    let m = i % period;
    if (m < 0) m += period;
    return m < n ? m : period - m;
}

/* 把 99 百分位 |x| 归一化到 1，并 clip 到 [-1, 1] */
function normalizeUnit(arr) {
    const p99 = percentileAbs(arr, 99);
    if (p99 < 1e-8) return arr;
    const s = 1.0 / p99;
    for (let i = 0; i < arr.length; i++) {
        let v = arr[i] * s;
        if (v < -1) v = -1;
        else if (v > 1) v = 1;
        arr[i] = v;
    }
    return arr;
}

/**
 * @param {Float32Array} theta      切线方向 (rad)
 * @param {Float32Array} coherence  各向异性强度 [0, 1]
 * @param {number} W
 * @param {number} H
 * @param {object} opts
 * @param {number} opts.length              streamline 长度 (px)
 * @param {number} opts.thickness           perpendicular 厚度 (px)
 * @param {number} [opts.seed=0]
 * @param {number} [opts.directionStrength=1.0]
 * @param {number} [opts.isoWeight=1.0]
 * @returns {Float32Array}                  stroke field, 大致 [-1, 1]，均值 ≈ 0
 */
export function directionalStrokeField(theta, coherence, W, H, opts) {
    const length = opts.length;
    const thickness = opts.thickness;
    const seed = opts.seed ?? 0;
    const directionStrength = opts.directionStrength ?? 1.0;
    const isoWeight = opts.isoWeight ?? 1.0;

    if (theta.length !== W * H || coherence.length !== W * H) {
        throw new Error('theta / coherence size mismatch with W*H');
    }

    const N = W * H;

    // 1. 白噪声
    const noise = standardNormalArray(N, seed);

    // 2. dx, dy
    const dx = new Float32Array(N);
    const dy = new Float32Array(N);
    for (let i = 0; i < N; i++) {
        dx[i] = Math.cos(theta[i]);
        dy[i] = Math.sin(theta[i]);
    }

    // 3. Streamline integration
    let stroked;
    if (length >= 1.0) {
        let nSteps = Math.max(3, Math.round(length));
        if (nSteps % 2 === 0) nSteps += 1;
        const half = (nSteps - 1) >> 1;
        const invSteps = 1.0 / nSteps;
        stroked = new Float32Array(N);
        for (let y = 0; y < H; y++) {
            for (let x = 0; x < W; x++) {
                const i = y * W + x;
                const dxi = dx[i];
                const dyi = dy[i];
                let acc = 0.0;
                for (let k = -half; k <= half; k++) {
                    const sy = y + k * dyi;
                    const sx = x + k * dxi;
                    // inline bilinear sample with reflect101
                    const y0f = Math.floor(sy);
                    const x0f = Math.floor(sx);
                    const fy = sy - y0f;
                    const fx = sx - x0f;
                    const y0 = reflect101(y0f, H);
                    const y1 = reflect101(y0f + 1, H);
                    const x0 = reflect101(x0f, W);
                    const x1 = reflect101(x0f + 1, W);
                    const v00 = noise[y0 * W + x0];
                    const v01 = noise[y0 * W + x1];
                    const v10 = noise[y1 * W + x0];
                    const v11 = noise[y1 * W + x1];
                    const top = v00 * (1 - fx) + v01 * fx;
                    const bot = v10 * (1 - fx) + v11 * fx;
                    acc += top * (1 - fy) + bot * fy;
                }
                stroked[i] = acc * invSteps;
            }
        }
    } else {
        stroked = noise.slice();
    }

    // 4. Perpendicular Gaussian thickness blur
    if (thickness >= 1.0) {
        let nThick = Math.max(3, Math.round(thickness));
        if (nThick % 2 === 0) nThick += 1;
        const thalf = (nThick - 1) >> 1;
        const sigmaT = Math.max(thickness / 2.0, 0.5);
        const ks = new Float32Array(nThick);
        const weights = new Float32Array(nThick);
        let wsum = 0.0;
        for (let i = 0; i < nThick; i++) {
            const k = i - thalf;
            ks[i] = k;
            const w = Math.exp(-0.5 * (k / sigmaT) * (k / sigmaT));
            weights[i] = w;
            wsum += w;
        }
        for (let i = 0; i < nThick; i++) weights[i] /= wsum;

        const thickened = new Float32Array(N);
        for (let y = 0; y < H; y++) {
            for (let x = 0; x < W; x++) {
                const i = y * W + x;
                const dxi = dx[i];
                const dyi = dy[i];
                let acc = 0.0;
                for (let kk = 0; kk < nThick; kk++) {
                    const k = ks[kk];
                    // 垂直方向旋转 90°: (-sin θ, cos θ) = (-dy, dx)
                    const sy = y + k * dxi;       // perp y = +cos θ = +dx
                    const sx = x + k * (-dyi);    // perp x = -sin θ = -dy
                    const y0f = Math.floor(sy);
                    const x0f = Math.floor(sx);
                    const fy = sy - y0f;
                    const fx = sx - x0f;
                    const y0 = reflect101(y0f, H);
                    const y1 = reflect101(y0f + 1, H);
                    const x0 = reflect101(x0f, W);
                    const x1 = reflect101(x0f + 1, W);
                    const v00 = stroked[y0 * W + x0];
                    const v01 = stroked[y0 * W + x1];
                    const v10 = stroked[y1 * W + x0];
                    const v11 = stroked[y1 * W + x1];
                    const top = v00 * (1 - fx) + v01 * fx;
                    const bot = v10 * (1 - fx) + v11 * fx;
                    acc += weights[kk] * (top * (1 - fy) + bot * fy);
                }
                thickened[i] = acc;
            }
        }
        stroked = thickened;
    }

    // 5. 归一化两路
    const strokedN = normalizeUnit(stroked);
    const isoN = normalizeUnit(noise.slice());

    // 6. 按 coherence 混合
    const out = new Float32Array(N);
    for (let i = 0; i < N; i++) {
        let b = coherence[i] * directionStrength;
        if (b < 0) b = 0;
        else if (b > 1) b = 1;
        out[i] = b * strokedN[i] + (1 - b) * isoN[i] * isoWeight;
    }
    return out;
}
